packages/context/src/context.ts

```diff
@@ -797,21 +799,27 @@ export class Context extends EventEmitter {
   ): ValueOrPromise<ValueType | undefined> {
     const {key, propertyPath} = BindingKey.parseKeyWithPath(keyWithPath);
 
-    optionsOrSession = asResolutionOptions(optionsOrSession);
-
-    const binding = this.getBinding<ValueType>(key, optionsOrSession);
-    if (binding == null) return undefined;
+    const options = asResolutionOptions(optionsOrSession);
+
+    let binding: Binding<ValueType> | undefined;
+    try {
+      binding = this.getBinding<ValueType>(key, options);
+      if (binding == null) return undefined;
+    } catch (err) {
+      throwResolutionError(err, {
```

Comment:
I still don't understand why we have to throw an error (deep) inside of `getBinding` and then catch & process it here. 

IIUC, there are only two situations where `getBinding` throws an error:
1. The binding key is invalid (it contains a property separator). https://github.com/strongloop/loopback-next/blob/e5e3d193a182b364a0877dd223be7a0d936a91cd/packages/context/src/binding-key.ts#L73-L83
2. The binding key was not found and `options.optional` is not set to `true`.

If my assessment is correct, then your proposal is removing just the one stack frame for `this.getBinding`, is it really that much of an improvement?

Setting the question aside, I think the implementation would be much simpler if this code invoked `getBinding` in a non-throwing mode and dealt with error conditions explicitly. 

Personally, I'd introduce `findBinding` that returns `undefined` when the binding was not found.

```ts
class Context {
  // ...
  getBinding<ValueType>(
    key: BindingAddress<ValueType>,
    options?: {optional?: boolean},
  ): Binding<ValueType> | undefined {
    const binding = this.findBinding(key, options);
    if (!binding && !options?.optional) {
      throw new Error(
      `The key '${key}' is not bound to any value in context ${this.name}`,
      );
    }
    return binding;
  }

  findBinding<ValueType>(
    key: BindingAddress<ValueType>,
    options?: {optional?: boolean},
  ): Binding<ValueType> | undefined {
    // the original implementation of `getBinding` goes here,
    // minus the part throwing an error
  }
}
```

Then you can rework the code in `getValueOrPromise` as follows:

```ts
    const binding = this.findBinding<ValueType>(key, options);
    if (!binding && !options?.optional) {
       throwResolutionError(/*...*/);
    }
```

Maybe that's not really needed and we just need to rework `getBinding` to throw a ResolutionError instead of plain Error?